### Today we are going to perform the simple classification of the amazon reviews' sentiment.

### Please, download the dataset amazon_baby.csv.

In [394]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import string
from sklearn.linear_model import LogisticRegression

def remove_punctuation(text):
    import string
    translator = str.maketrans('', '', string.punctuation)
    return text.translate(translator)

baby_df = pd.read_csv('amazon_baby.csv')
baby_df.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  
0  These flannel wipes are OK, but in my opinion ...       3  
1  it came early and was not disappointed. i love...       5  
2  Very soft and comfortable and warmer than it l...       5  
3  This is a product well worth the purchase.  I ...       5  
4  All of my kids have cried non-stop when I trie...       5

## Exercise 1 (data preparation)
a) Remove punctuation from reviews using the given function.   
b) Replace all missing (nan) revies with empty "" string.  
c) Drop all the entries with rating = 3, as they have neutral sentiment.   
d) Set all positive ($\geq$4) ratings to 1 and negative($\leq$2) to -1.

##Comment

First step for prepearing data it's remove punctuation from reviews. It will be better for our model. I did by using regex pattern for punctuations and replace it by empty string. And showed short test for index = 3

In [357]:
#a)
baby_df['review'] = baby_df['review'].str.replace(r'[^\w\s]+', '')
#short test: 
baby_df["review"][4] == 'All of my kids have cried nonstop when I tried to ween them off their pacifier until I found Thumbuddy To Loves Binky Fairy Puppet  It is an easy way to work with your kids to allow them to understand where their pacifier is going and help them part from itThis is a must buy book and a great gift for expecting parents  You will save them soo many headachesThanks for this book  You all rock'
remove_punctuation(baby_df["review"][4]) == 'All of my kids have cried nonstop when I tried to ween them off their pacifier until I found Thumbuddy To Loves Binky Fairy Puppet  It is an easy way to work with your kids to allow them to understand where their pacifier is going and help them part from itThis is a must buy book and a great gift for expecting parents  You will save them soo many headachesThanks for this book  You all rock'

<ipython-input-357-9f11ee4d0f17>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  baby_df['review'] = baby_df['review'].str.replace(r'[^\w\s]+', '')


True

In [358]:
baby_df["review"][3] == 'This is a product well worth the purchase  I have not found anything else like this and it is a positive ingenious approach to losing the binky  What I love most about this product is how much ownership my daughter has in getting rid of the binky  She is so proud of herself and loves her little fairy  I love the artwork the chart in the back and the clever approach of this tool'

True

## Comment

Here I replaced all missing (nan) revies with empty "" string.

In [359]:
#b)
baby_df['review'] = baby_df['review'].fillna("")
#short test:
baby_df["review"][38] == baby_df["review"][38]

True

In [360]:
baby_df["review"][38]

''

## Comment

Here I droped all the entries with rating = 3, as they have neutral sentiment.

In [361]:
#c)
baby_df = baby_df.drop(baby_df[baby_df.rating == 3].index)
#short test:
sum(baby_df["rating"] == 3)

0

## Comment

 I set all positive ( ≥ 4) ratings to 1 and negative( ≤ 2) to -1.
 I did it by using mask function and I showed for different index

In [362]:
#d) 
baby_df['rating'].mask(baby_df['rating'] <= 2, -1, inplace=True)
baby_df['rating'].mask(baby_df['rating'] >= 4, 1, inplace=True)
#baby_df.loc[baby_df["rating"] <= 2, "rating"] = -1
#short test:
sum(baby_df["rating"]**2 != 1)

0

In [363]:
baby_df.head(25)

name  \
1                               Planetwise Wipe Pouch   
2                 Annas Dream Full Quilt with 2 Shams   
3   Stop Pacifier Sucking without tears with Thumb...   
4   Stop Pacifier Sucking without tears with Thumb...   
5   Stop Pacifier Sucking without tears with Thumb...   
6             A Tale of Baby's Days with Peter Rabbit   
7   Baby Tracker&reg; - Daily Childcare Journal, S...   
8   Baby Tracker&reg; - Daily Childcare Journal, S...   
9   Baby Tracker&reg; - Daily Childcare Journal, S...   
10  Baby Tracker&reg; - Daily Childcare Journal, S...   
11  Baby Tracker&reg; - Daily Childcare Journal, S...   
12  Baby Tracker&reg; - Daily Childcare Journal, S...   
14     Nature's Lullabies First Year Sticker Calendar   
15     Nature's Lullabies First Year Sticker Calendar   
16     Nature's Lullabies First Year Sticker Calendar   
17    Nature's Lullabies Second Year Sticker Calendar   
18    Nature's Lullabies Second Year Sticker Calendar   
19    Nature's Lullabies Second Year Sticker Calendar   
20    Nature's Lullabies Second Year Sticker Calendar   
21    Nature's Lullabies Second Year Sticker Calendar   
22    Nature's Lullabies Second Year Sticker Calendar   
24    Nature's Lullabies Second Year Sticker Calendar   
25    Nature's Lullabies Second Year Sticker Calendar   
26                       Baby's First Journal - Green   
28                        Lamaze Peekaboo, I Love You   

                                               review  rating  
1   it came early and was not disappointed i love ...       1  
2   Very soft and comfortable and warmer than it l...       1  
3   This is a product well worth the purchase  I h...       1  
4   All of my kids have cried nonstop when I tried...       1  
5   When the Binky Fairy came to our house we didn...       1  
6   Lovely book its bound tightly so you may not b...       1  
7   Perfect for new parents We were able to keep t...       1  
8   A friend of mine pinned this product on Pinter...       1  
9   This has been an easy way for my nanny to reco...       1  
10  I love this journal and our nanny uses it ever...       1  
11  This book is perfect  Im a first time new mom ...       1  
12  I originally just gave the nanny a pad of pape...       1  
14  Space for monthly photos info and a lot of use...       1  
15  I bought this calender for myself for my secon...       1  
16  I love this little calender you can keep track...       1  
17  This was the only calender I could find for th...       1  
18  I completed a calendar for my sons first year ...       1  
19  We wanted to get something to keep track of ou...       1  
20  I had a hard time finding a second year calend...       1  
21  I only purchased a secondyear calendar for my ...      -1  
22  I LOVE this calendar for recording events of m...       1  
24  Wife loves this calender Comes with a lot of s...       1  
25  My daughter had her 1st baby over a year ago S...       1  
26  Extremely useful As a new mom tired and inexpe...       1  
28  One of babys first and favorite books and it i...       1

## CountVectorizer
In order to analyze strings, we need to assign them numerical values. We will use one of the simplest string representation, which transforms strings into the $n$ dimensional vectors. The number of dimensions will be the size of our dictionary, and then the values of the vector will represent the number of appereances of the given word in the sentence.

In [364]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
reviews_train_example = ["We like apples",
                   "We hate oranges",
                   "I adore bananas",
                   "We like like apples and oranges",
                   "They dislike bananas"]

X_train_example = vectorizer.fit_transform(reviews_train_example)

print(vectorizer.get_feature_names())
print(X_train_example.todense())



['adore', 'and', 'apples', 'bananas', 'dislike', 'hate', 'like', 'oranges', 'they', 'we']
[[0 0 1 0 0 0 1 0 0 1]
 [0 0 0 0 0 1 0 1 0 1]
 [1 0 0 1 0 0 0 0 0 0]
 [0 1 1 0 0 0 2 1 0 1]
 [0 0 0 1 1 0 0 0 1 0]]


/usr/local/lib/python3.8/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [365]:
reviews_test_example = ["They like bananas",
                   "We hate oranges bananas and apples",
                   "We love bananas"] #New word!

X_test_example = vectorizer.transform(reviews_test_example)

print(X_test_example.todense())

[[0 0 0 1 0 0 1 0 1 0]
 [0 1 1 1 0 1 0 1 0 1]
 [0 0 0 1 0 0 0 0 0 1]]


We should acknowledge few facts. Firstly, CountVectorizer does not take order into account. Secondly, it ignores one-letter words (this can be changed during initialization). Finally, for test values, CountVectorizer ignores words which are not in it's dictionary.

## Exercise 2 
a) Split dataset into training and test sets.     
b) Transform reviews into vectors using CountVectorizer. 

## Comment

I splitted dataset and transformed reviews into vectors using CountVectorizer.

In [366]:
#a)
from sklearn.model_selection import train_test_split
reviews_train_df, reviews_test_df = train_test_split(baby_df, test_size=0.4)

In [367]:
X_train = reviews_train_df['review']
y_train = reviews_train_df['rating']

In [368]:
X_test = reviews_test_df['review']
y_test = reviews_test_df['rating']

In [369]:
#b)
from sklearn.feature_extraction.text import CountVectorizer
df_vectorizer = CountVectorizer()

X_train = df_vectorizer.fit_transform(X_train)
X_test = df_vectorizer.transform(X_test)

## Exercise 3 
a) Train LogisticRegression model on training data (reviews processed with CountVectorizer, ratings as they were).   
b) Print 10 most positive and 10 most negative words.

## Comment

I trained Logistic Regression model on training data. For X I used reviews processed with Count Vectorizer. For y I used ratings as they from begginig. 

Also I changed max itterations for model. Because of default 100 was not enough and throwed warning about it.

And I used magical function %time gives us a single run of a function

In [370]:
#a)
model = LogisticRegression(max_iter=1000)
%time model.fit(X_train,y_train)

CPU times: user 1min 29s, sys: 1min 40s, total: 3min 9s
Wall time: 1min 40s


LogisticRegression(max_iter=1000)

Here I printed TOP 10 Positive and Negative words by using model.coef_.

I sorted values and took 10 values, after this I found index and mapped to words using vectorizer.get_feature_names_out function.<p>
model.coef_ - Coefficient of the features in the decision function.

In [371]:
#b)
ten_positive = sorted(model.coef_[0])[-10:]
ten_negative = sorted(model.coef_[0])[:10]

print('Positive')
for item in reversed(ten_positive):
    index = np.where(model.coef_[0] == item)
    print(f'coef: {model.coef_[0][index]}     word: {df_vectorizer.get_feature_names_out()[index]}')

print('\nNegative')
for item in ten_negative:
    index = np.where(model.coef_[0] == item)
    print(f'coef: {model.coef_[0][index]}     word: {df_vectorizer.get_feature_names_out()[index]}')

#hint: model.coef_, vectorizer.get_feature_names()

Positive
coef: [2.27168695]     word: ['wonderfully']
coef: [2.20433514]     word: ['lifesaver']
coef: [2.1550751]     word: ['highly']
coef: [2.01489667]     word: ['breeze']
coef: [2.00366026]     word: ['worry']
coef: [1.98545888]     word: ['amazed']
coef: [1.96734841]     word: ['fancy']
coef: [1.89767119]     word: ['glad']
coef: [1.89459578]     word: ['saves']
coef: [1.88394568]     word: ['tad']

Negative
coef: [-3.031933]     word: ['dissapointed']
coef: [-2.782497]     word: ['disappointing']
coef: [-2.64621289]     word: ['worst']
coef: [-2.50487479]     word: ['unusable']
coef: [-2.40662627]     word: ['useless']
coef: [-2.37059925]     word: ['worthless']
coef: [-2.36756834]     word: ['poorly']
coef: [-2.29004876]     word: ['theory']
coef: [-2.27939768]     word: ['poor']
coef: [-2.23256142]     word: ['disappointed']


## Exercise 4 
a) Predict the sentiment of test data reviews.   
b) Predict the sentiment of test data reviews in terms of probability.   
c) Find five most positive and most negative reviews.   
d) Calculate the accuracy of predictions.

## Comment
I predicted the sentiment of test data reviews.
I used predict function which predict class labels for samples in X.<p>
And returns vector containing the class labels for each sample.

In [372]:
#a)
model.predict(X_test)

array([ 1,  1,  1, ...,  1,  1, -1])

I predicted the sentiment of test data reviews in terms of probability.<p>
I used predict_proba function which using for probability estimates.<p>
And returns the probability of the sample for each class in the model, where classes are ordered as they are in self.classes_.<p>
For our example order of classes will be [-1,1]. It means thar first column will be for negative reviews and second column for positive values

In [373]:
#b)
pr_proba = model.predict_proba(X_test)
print(pr_proba)
#hint: model.predict_proba()

[[0.03147975 0.96852025]
 [0.00188307 0.99811693]
 [0.00292813 0.99707187]
 ...
 [0.3447533  0.6552467 ]
 [0.00253843 0.99746157]
 [0.9721434  0.0278566 ]]


Here I found TOP 5 reviews. The same steps how I did for TOP 10 words.

In [374]:
#c) 
ids_neg = np.arange(pr_proba[:, 0].shape[0])
res_neg = sorted(ids_neg, key=lambda x: pr_proba[:, 0][x])

ids_pos = np.arange(pr_proba[:, 1].shape[0])
res_pos = sorted(ids_pos, key=lambda x: pr_proba[:, 1][x])

In [375]:
print('Negative Review:')
for index in reversed(res_neg[:5]):
    print(f"{reviews_test_df['review'].iloc[index]} \n")

Negative Review:
I have been using this version of the Britax Marathon 70 for 2 years  I originally bought it just after my daughter turned a year old and we needed to move her out of her bucket style rearfacing seat After a lot of research I chose the Britax Marathon for its excellent safety standards and because I thought it would fit nicely in my carMy vehicle setup 2004 Pontiac VibeI felt that the Marathon fit very nicely rearfacing in my car Due to the split seats in the back I kept the car seat on the driver side of the car I am not a very tall person 5 5 so I was able to properly install the seat rearfacing without feeling squished Inn fact I had the seat where I had always had it My husband who is 5 9 was also able to comfortably able to drive the car with the Marathon rearfacing although he was a tad squishedI kept my daughter rearfacing until she was almost 3 years old She never once complained of any discomfort Yes she had to bend her legs a little but kids are pretty flexib

In [376]:
print('Positive Review:')
for index in reversed(res_pos[-5:]):
    print(f"{reviews_test_df['review'].iloc[index]} \n")

#hint: use the results of b)

Positive Review:
We live an active life style and are either traveling on weekends or are away from home with our two year old twins doing activities As they have both been potty trained I felt that it was important to ensure that they have access to a potty at all times while we are outside the home After visiting numerous public rest rooms and repeatedly hearing my twin say mommy clean it each time we entered a public restroom I decided to search for a portable potty  When it arrived it was significantly smaller than I expected and I appreciated that it came with a storage bag  which I have used for both the potty and the seat insertsThis potty was first used on a 35 hour trip to another state and we were all so thankful that it was in the car Like others I didnt get out it worked initially since I didnt read the instructions and held it upright while my child when to the bathroom Later I discovered that you need to push in the legs to ensure that the potty is sturdy and can support 

Here I calculate the accuracy of predictions. Function score returns the mean accuracy on the given test data and labels.

In [377]:
#d) 
model.score(X_test, y_test)

0.9307506634083447

## Exercise 5
In this exercise we will limit the dictionary of CountVectorizer to the set of significant words, defined below.


a) Redo exercises 2-5 using limited dictionary.   
b) Check the impact of all the words from the dictionary.   
c) Compare accuracy of predictions and the time of evaluation.

## Comment

Here I repeated exercises 2-5 using limited dictionary. I dit it by specifying vocabulary(using significant_words) in CountVectorizer. And also calcutated time of model.fit function for this model.

We can notice that using significant words we get others reviews. When I read them for me they were very similar. I think it is because we have less words for trainning. The results of model and impact I will disccuss in next points.

In [378]:
significant_words = ['love','great','easy','old','little','perfect','loves','well','able','car','broke','less','even','waste','disappointed','work','product','money','would','return']

In [379]:
#a)
from sklearn.model_selection import train_test_split
reviews_train_df, reviews_test_df = train_test_split(baby_df, test_size=0.4)

In [380]:
X_train = reviews_train_df['review']
y_train = reviews_train_df['rating']

In [381]:
X_test = reviews_test_df['review']
y_test = reviews_test_df['rating']

In [382]:
from sklearn.feature_extraction.text import CountVectorizer
df_vectorizer = CountVectorizer(vocabulary=significant_words)

X_train = df_vectorizer.fit_transform(X_train)
X_test = df_vectorizer.transform(X_test)

In [393]:
model = LogisticRegression(max_iter=1000)
%time model.fit(X_train, y_train)

CPU times: user 521 ms, sys: 4.5 ms, total: 526 ms
Wall time: 547 ms


LogisticRegression(max_iter=1000)

In [384]:
ten_positive = sorted(model.coef_[0])[-10:]
ten_negative = sorted(model.coef_[0])[:10]

print('Positive')
for item in reversed(ten_positive):
    index = np.where(model.coef_[0] == item)
    print(f'coef: {model.coef_[0][index]}     word: {df_vectorizer.get_feature_names_out()[index]}')

print('\nNegative')
for item in ten_negative:
    index = np.where(model.coef_[0] == item)
    print(f'coef: {model.coef_[0][index]}     word: {df_vectorizer.get_feature_names_out()[index]}')

Positive
coef: [1.7782034]     word: ['loves']
coef: [1.46231894]     word: ['perfect']
coef: [1.38115099]     word: ['love']
coef: [1.17731017]     word: ['easy']
coef: [0.95897369]     word: ['great']
coef: [0.50269334]     word: ['little']
coef: [0.46837384]     word: ['well']
coef: [0.20695297]     word: ['able']
coef: [0.04801978]     word: ['old']
coef: [0.03910323]     word: ['car']

Negative
coef: [-2.4047427]     word: ['disappointed']
coef: [-2.15554998]     word: ['return']
coef: [-1.94311948]     word: ['waste']
coef: [-1.66275156]     word: ['broke']
coef: [-0.95430416]     word: ['money']
coef: [-0.61905652]     word: ['work']
coef: [-0.53301052]     word: ['even']
coef: [-0.33390596]     word: ['would']
coef: [-0.30499985]     word: ['product']
coef: [-0.22761133]     word: ['less']


In [385]:
model.predict(X_test)

array([1, 1, 1, ..., 1, 1, 1])

In [386]:
pr_proba = model.predict_proba(X_test)
print(pr_proba)

[[0.09383961 0.90616039]
 [0.15118412 0.84881588]
 [0.07684791 0.92315209]
 ...
 [0.21271191 0.78728809]
 [0.02344678 0.97655322]
 [0.12423148 0.87576852]]


In [387]:
ids_neg = np.arange(pr_proba[:, 0].shape[0])
res_neg = sorted(ids_neg, key=lambda x: pr_proba[:, 0][x])

ids_pos = np.arange(pr_proba[:, 1].shape[0])
res_pos = sorted(ids_pos, key=lambda x: pr_proba[:, 1][x])

In [388]:
print('Negative Review:')
for index in reversed(res_neg[:5]):
    print(f"{reviews_test_df['review'].iloc[index]} \n")

Negative Review:
We bought this stroller after selling our beloved BOB rev on craigslist We used the BOB for 9 months for my son but it just wasnt practical I dont jogrun it didnt have a big basket and was very bulky to take into stores quickly However I did love how it unfolded easily but it was heavy to fold up and lift into my small trunk myself Overall I didnt realize what Id need in a stroller until AFTER I had my son Live  learn We did love how easily the BOB would go over pretty much anything Nevertheless we sold it and after extensive research on strollers we decided it was between the uppababy brand because of the large baskets OR the city mini GT because of its easy fold up design After looking over both strollers I decided on the uppababy cruz because of a few main factors It SITS UP I cant tell you how much my son hates being reclined when he is just riding in the stroller and not napping The BOB and the City Mini had a slight recline and he always tried sitting more uprigh

In [389]:
print('Positive Review:')
for index in reversed(res_pos[-5:]):
    print(f"{reviews_test_df['review'].iloc[index]} \n")

Positive Review:
Check and recheck the KTan for size issues before letting your guard down and wearing NO ISSUES OCCURED however I thought I was good to go and then realized that this large was different than the large I had been given before it Odd but fabric This makes the moby wrap appear like a midevil torture devise and bonus you wont have to worry about some BabyWearing freak watching you in the parking lot at the mall as you struggle to turn a bolt of fabric into what the KTan has simplifiedseriously Dad approved and if the moby is your thing thats cool just perhaps realize that its not as easy as everyone thinks to wrap that amount of fabric around yourself evenly snug comfortable and without mummifying yourself and lord help you your baby  Another bonus to the KTAN issue with baby takes only a second to move her out of the KTAN but in a moby We thought for the first two days after returning the moby 34have you seen Katy34love it love it love it love it love it love it love it 

In [390]:
model.score(X_test, y_test)

0.866433786599901

 ## Comment
 Here checked the impact of all the words from the dictionary. I understood it like printing coefficients for words. And we can see the word with the largest coefficient and with the smallest.

In [391]:
#b)
ids_c =  np.arange(model.coef_[0].shape[0])
res_c = sorted(ids_c, key=lambda x: model.coef_[0][x],reverse=True)

print('Impact:')
for index in res_c:
    print(f'coef: {model.coef_[0][index]}     word: {df_vectorizer.get_feature_names_out()[index]}')

Impact:
coef: 1.7782033978422536     word: loves
coef: 1.4623189427749166     word: perfect
coef: 1.381150987622393     word: love
coef: 1.1773101655818772     word: easy
coef: 0.9589736933476037     word: great
coef: 0.5026933378950218     word: little
coef: 0.4683738388298545     word: well
coef: 0.20695296669458307     word: able
coef: 0.048019783051560294     word: old
coef: 0.03910323416986804     word: car
coef: -0.2276113257969504     word: less
coef: -0.3049998483495305     word: product
coef: -0.3339059558741242     word: would
coef: -0.5330105223344395     word: even
coef: -0.6190565166327158     word: work
coef: -0.9543041592237376     word: money
coef: -1.6627515600010505     word: broke
coef: -1.9431194802202838     word: waste
coef: -2.1555499769131585     word: return
coef: -2.4047426957775953     word: disappointed


## Comment

Compare accuracy of predictions and the time of evaluation.

The accuracy of our first model was 0.9307506634083447.<p>
Here we have 0.866433786599901.<p>

It's a big difference and reason of it was in amount of our train data and lenght of our vacabulary for learning. Using limited dictionary model will have smaller accuracy


For time of evaluation:<p>

 - Time for **first model** where we have all words:

    CPU times: user 1min 30s, sys: 1min 46s, total: 3min 17s<p>
    Wall time: 1min 44s

 - Time for **second model** where we have only 20 words: 
    CPU times: user 521 ms, sys: 4.5 ms, total: 526 ms<p>
    Wall time: 547 ms

So we can see that the more words and "opportunity" for learning we have, the longer it takes to learn.

In [392]:
#c)
model.score(X_test, y_test)
#hint: %time, %timeit

0.866433786599901